### PR TITLE
storage: invoke a separate raft RPC on snapshot

### LINKED ISSUE #3013: storage: snapshots transport require very high demands on the network
Now default range size is 64MB. If there is a snapshot transporting, it will occupy rpc channel for some time. Rpc and raft heartbeat will be blocked. 
Default raft heartbeat interval is 3s. This means that it must transfer 64MB or more snapshot data in the 3s, which the network requirements are very high. Otherwise raft will re-election. 
Especially for remote nodes, heartbeat timeout situation should be easy to happen.

### PR DESCRIPTION
**Pull request by tamird**
Fixes #3013.

I suppose I'd call this a WIP, since it's untested and currently just spawns an untracked goroutine. Should we track said goroutine?

I'm not sure how to test this except by inducing message reordering, which seems sketchy.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5276)

### CODE REVIEW

File: storage/raft_transport.go [hunk 1]
```diff
@@ -225,9 +229,25 @@ func (t *RaftTransport) processQueue(nodeID roachpb.NodeID) {
...
 		case req := <-ch:
-			if err := stream.Send(req); err != nil {
-				log.Error(err)
-				return
+			if len(req.Message.Snapshot.Data) > 0 {
+				go func() {
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+					stream, err := makeStream(ctx)
+					if err != nil {
```
Comment (BramGruneir) [inline storage/raft_transport.go]: if err != nil && log.V(1) {

**Comment by BramGruneir** [comment]
LGTM, but you do need to figure out a way to test it.
Maybe move the for loop into a separate function?  Mock out the send.  Ick.

**Comment by tamird** [comment]
Some strange is going on here. This change increases the runtime of package `kv`'s tests by 12 seconds.

**Comment by bdarnell** [comment]
For testing, we could add an assertion to `RaftTransport.RaftMessage` that identifies the stream as "snapshot" or "other" (based on the first message) and fails if we get a message on the wrong stream. Or we could define a separate rpc on the `MultiRaft` service and disallow any messages that come in on the wrong interface.  (that would be a good long-term move since it would let us eventually move away from holding the entire snapshot in memory at once)

I don't have any idea why this would slow the tests down so much. 

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 237 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KCvzE-3bgn_CAQRWu2p-r1-237:-KD-OUDEF_Hc4fLz5uZ_:1587893135) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4346b107a5558296e95843aa7ee3affe679e28d/storage/raft_transport.go#L237)):*
Why? we use `log.Error` without a `V(1)` guard for the non-snapshot `stream.Send`. 

---

*[storage/raft_transport.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD-Jem7pjuTG11n10RC:-KD-Jem7pjuTG11n10RD:-497357820) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc5998a3a71b96bc603baa4decbbe6017cf7cc79/storage/raft_transport.go#L228)):*
The check should probably be `req.Message.Type == raftpb.MsgSnap`

---

*[storage/raft_transport.go, line 229 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD-Jk9ssN_SvyVpGiwI:-KD-Jk9ssN_SvyVpGiwJ:-490751924) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc5998a3a71b96bc603baa4decbbe6017cf7cc79/storage/raft_transport.go#L229)):*
I think this should be a `RunAsyncTask`. I might do everything but the `stream.Send` before going into the task; only the actual message send needs to be async.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-:-KD-KotFSI7Fw3I7AXzF:1218150814)*

**Comment by tamird** [comment]
Updated this and added a test, and the test slowdown seems to have gone away. Goblins? Gremlins? I'm not sure.

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/raft_transport.go, line 237 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KCvzE-3bgn_CAQRWu2p-r1-237:-KD-iSKp67gpozrxNavM:-1207971761) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4346b107a5558296e95843aa7ee3affe679e28d/storage/raft_transport.go#L237)):*
This has changed a bit.

---

*[storage/raft_transport.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD-Jem7pjuTG11n10RC:-KD-hyh_Y0S9FzD8KGu_:1320986700) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc5998a3a71b96bc603baa4decbbe6017cf7cc79/storage/raft_transport.go#L228)):*
Yeah, made that change locally.

---

*[storage/raft_transport.go, line 229 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD-Jk9ssN_SvyVpGiwI:-KD-iOirzkncaPDkbG37:1336915767) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc5998a3a71b96bc603baa4decbbe6017cf7cc79/storage/raft_transport.go#L229)):*
OK, done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-:-KD03avWBWS3DnaQ8AhX:-1994385342)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 239 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD3OK-4T28xTqCRo0x-:-KD3OK-4T28xTqCRo0x0:1076603030) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6ea93df3a1aeb75b82b2f43f308071d88cf7fb6/storage/raft_transport.go#L239)):*
I think the error should be logged at V(0) (same as below) and the success message at V(1).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-:-KD3Pg4-l1Nn1mzbwGSy:676834233)*

**Comment by tamird** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/raft_transport.go, line 239 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276#-KD3OK-4T28xTqCRo0x-:-KD3xSppRMwRNM9lq1iL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6ea93df3a1aeb75b82b2f43f308071d88cf7fb6/storage/raft_transport.go#L239)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5276)*

**Pull request merged**